### PR TITLE
refactor: inline `quicksink` crate

### LINKED ISSUE #4705: Adopt quicksink
### Description

libp2p should inline quicksink into the rust-libp2p monorepo. libp2p is a notable consumer of quicksink with overlapping developers from the current organization responsible for it.

### Motivation

Despite libp2p's active usage of quicksink, it hasn't been maintained since 2022 and has dated dependencies. While I would simply make a PR to quicksink to bump said dependencies, I can't as the repository is archived.

I did attempt to remove the reliance on quicksink as discussed a year and a half ago in https://github.com/libp2p/rust-libp2p/issues/2653#issuecomment-1133934276. Unfortunately, offering the `Sink` API around something without the `Sink` API *requires* effectively the entire quicksink library of code. Accordingly, there's no benefit to removing it vs adopting it.

The alternative would be to use future's `unfold`, dropping the ability to `flush` *and* deferring the `close` operation to `Drop`. While I can't claim experience with this level, that seemed unwise.

### Requirements

Potentially the consent of the prior organization to accept the quicksink crate unless libp2p wants to inline it as a file within libp2p-websocket (the one consumer)/rename it to `quicksink2`.

### Open questions

_No response_

### Are you planning to do it yourself in a pull request ?

Maybe

### PR DESCRIPTION
**Pull request by kayabaNerve**
## Description

Adds the quicksink crate, which is offered under the MIT or Apache 2 license, as a module. This enables maintenance since quicksink itself has been archived for a year in a half.

Resolves #4705.

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

### CODE REVIEW

**Comment by kayabaNerve** [comment]
Feel free to say this also shouldn't have a changelog entry.

File: transports/websocket/src/quicksink.rs [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+            let values = vec![Ok(&b"hello\n"[..]), Ok(&b"world\n"[..])];
+            assert!(stream::iter(values).forward(sink).await.is_ok())
+        })
+    }
+
+    #[test]
+    fn replay() {
+        task::block_on(async {
+            #[allow(clippy::disallowed_methods)]
+            let (tx, rx) = mpsc::unbounded();
```
Comment (thomaseizinger) [inline transports/websocket/src/quicksink.rs]: Can we just use a bounded channel here?

File: transports/websocket/CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,6 @@
 ## 0.43.0 - unreleased
 
+- Removes dependency on the archived `quicksink` by inlining it as a module.
```
Comment (thomaseizinger) [inline transports/websocket/CHANGELOG.md]: As you predicted, I don't think we need to mention internals like this :)

```suggestion
```

**Review by thomaseizinger — APPROVED** [review]

File: transports/websocket/src/quicksink.rs (hunk 1 above)
Comment (kayabaNerve) [reply transports/websocket/src/quicksink.rs]: Considering it was in a test alone, this was the simplest way forward. I'll evaluate length of the test's runtime...

**Review by thomaseizinger — APPROVED** [review]
Nice, thank you!

**Review by mxinden — COMMENTED** [review]
:+1: for inlining as a module.

**Pull request merged**